### ISSUE TITLE
Design documentation of the new ISAAC

### ISSUE THREAD

**Issue by Charleslee522** [bosnet/sebak#201]
Document for explaining design of the new ISAAC consensus protocol

**Comment by cmcm2222** [comment]
I think yesterday you showed to us. right??

**Comment by Charleslee522** [comment]
yes but we need to improve the document :)

**Comment by cmcm2222** [comment]
@Charleslee522  Do you have any ideas for doc modifying?
Or If possible, you can give me simple flow diagram that wrote in hand. then I will do.

**Comment by sconeman** [comment]
@cmcm2222 @Charleslee522 
@spikeekips has changed a lot. Maybe we should listen to his explanation.

**Comment by Charleslee522** [comment]
@cmcm2222 @spikeekips @spikeekips @Geod24 @sconeman @kyehyukahn 
Please review this page https://github.com/bosnet/sebak/wiki/How-the-consensus-protocol-works and feel free to change it.

**Comment by sconeman** [comment]
@Charleslee522 So Does ISAAC have timers by the ballot state?

**Comment by Charleslee522** [comment]
@sconeman It should be, but not yet. Only TimeoutExpireRound(10 sec) is now.

**Comment by sconeman** [comment]
@Charleslee522 In `ACCEPT` → `ALL-CONFIRM` chapter,
> 1. Each node checks the block is valid.
I think it is more good that this part is moved to `SIGN`  → `ALL-CONFIRM` chapter.
Because the description of `SIGN` is below.
> Checked that the proposed block is valid by itself

**Comment by Charleslee522** [comment]
@sconeman You're right. Thanks :)

**Comment by Charleslee522** [comment]
@cmcm2222 @spikeekips @spikeekips @Geod24 @sconeman @kyehyukahn 
I changed [design doc](https://github.com/bosnet/sebak/wiki/How-the-consensus-protocol-works), does it make sence?
@spikeekips How about voting without voting hole? I can assume the design that each node only votes when it agree with the proposed block.

**Comment by Charleslee522** [comment]
@spikeekips 
@Geod24 has a question. Why does a block have only valid transactions? Let's talk about it in this issue by comment :)

**Comment by spikeekips** [comment]
> @Charleslee522 : @spikeekips How about voting without voting hole? I can assume the design that each node only votes when it agree with the proposed block.

If node does not agree to the proposed ballot, this node will broadcast the no voting ballot, so the other node can determine the proposed ballot is over threshold by yes voting, excluding no voting. If no voting is not broadcasted, the other nodes can not know.

**Comment by Charleslee522** [comment]
@spikeekips @Geod24 I misunderstood the voting ballot design.
Each node broadcasts the ballot with all transactions but only valid transactions are included.

**Comment by spikeekips** [comment]
@Charleslee522 @Geod24 Please describe more details :)

**Comment by Charleslee522** [comment]
@spikeekips @Geod24 Oh, I should change the sentence.
> Why does a block have only valid transactions?

Why does a voting ballot have whole transactions? Not only valid transactions?

**Comment by spikeekips** [comment]
> @Charleslee522 : Why does a voting ballot have valid and invalid transactions? Not only valid transactions?

We can think this situation,

1. 3 nodes: A, B, C
2. the proposer node, A proposed the proposed ballot
3. one node, B received the proposed ballot without problem,
4. but the other node, C failed to receive the proposed ballot with some reason
5. C received the voted ballot from B

If the voted ballot in step 5 does not have the original proposed ballot from A, the node, C can not validate the voted ballot of B. This kind of cases also will be applied in ACCEPT state.

**Comment by Geod24** [comment]
I'm not sure I follow. @Charleslee522 question was about transactions, not ballot.
But that seems to be a non-issue, since it's all come down from a misunderstanding, as far as I can see.

So let me rephrase the question: What does a node do when it receives an invalid transaction ?

**Comment by sconeman** [comment]
@spikeekips I did not understand well. I understand that your words are not about the `invalid transactions` about `missed proposed ballots`, right?

**Comment by spikeekips** [comment]
> @Geod24 What does a node do when it receives an invalid transaction ?

To clarify your question, 'invalid' transaction is not 'not well-formed' transaction. If the transaction,
* has bad signature
* unmarshal failed
* has under base fee
* zero amount

I call this kind of transaction is 'not-wellformed' transaction, not 'invalid' transaction. The invalid transaction means not-valid like,
* has unknown checkpoint
* the source account has under balance
* there is no target account in payment operation
* etc

**Comment by Geod24** [comment]
@spikeekips : Thanks for the clarification, that vocabulary makes sense.
So if a [malformed](https://www.merriam-webster.com/dictionary/malformed) (not well formed) transaction is received, they are ignored, which makes sense.

What about transactions which are well formed but not valid ?

**Comment by spikeekips** [comment]
@Geod24 Not valid transactions will be validated by proposer and only the valids will be in ‘ValidTransactions’ and invalids will be listed in ‘Transactions’ with valids.
